### PR TITLE
[WIP] Got typecheck tests running (#148)

### PR DESCRIPTION
**Pull request by hershd23**
## Description
Integrated typechecking code from PySyft 0.3.0 and got the typechecking tests running. Changes made to some other files are stashed as the other tests have not been updated (is an existing issue) 

## How has this been tested?
Used source and test code from PySyft 0.3.0 for typechecking and got the tests for typechecking running. 

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by hershd23** [comment]
For the ease of understanding for the reviewers, the following comments have been made in place of @type_hints above function declaration:-

1. **Find return type :  <variable name>** - if the return type isn't clear (E.g. for vectors). Will be clear once the numpy/tensor issue gets fixed
2. **Find argument type: <variable name>** - if one of the argument types isn't clear
3. **On @type_hints gives : <variable name> not defined, on importing leads to a circular import** -  if the typecheck wants an import but importing leads to circular imports
4. **Import inside: <variable name>** -  if an import has been made within a method to avoid circular imports, creates a situation where typechecking gives errors.

If you find more such errors written as comments, I'll include them here. Also if you know **the type for the variables in question** or **a workaround for any of these,** reach out so they can be fixed

**Comment by hershd23** [comment]
Errors of type 3 & 4 mentioned above have been resolved.

File: syfertext/doc.py [hunk 1]
```diff
@@ -94,6 +101,7 @@ def remove_attribute(self, name: str):
 
         delattr(self._, name)
 
+    # Find return type
```
Comment (Nilanshrajput) [inline syfertext/doc.py:104]: It should be bool or a string

File: syfertext/doc.py [hunk 2]
```diff
@@ -155,22 +164,26 @@ def __getitem__(self, key: Union[int, slice]) -> Union[Token, Span, int]:
...
             yield self[i]
 
     @property
-    def text(self):
+    @type_hints
+    def text(self) -> str:
         """Returns the text present in the doc with whitespaces"""
         return "".join(token.text_with_ws for token in self)
 
+    # Find Return Type: Vector
```
Comment (Nilanshrajput) [inline syfertext/doc.py:186]: when master is merged into this branch it would be `torch.Tensor` but currently its np.array so let's keep this on hold

File: syfertext/doc.py [hunk 3]
```diff
@@ -241,6 +257,7 @@ def similarity(self, other: "Doc") -> torch.Tensor:
 
         return sim
 
+    # Find Return Type: Vector
```
Comment (Nilanshrajput) [inline syfertext/doc.py:260]: same as in case of vector, methods handling vectors could be updated after we merge master into this branch.

File: syfertext/language.py [hunk 4]
```diff
@@ -122,7 +126,8 @@ def set_vocab(self, vocab: Vocab, access: Set[str] = None) -> None:
         # Save the state in the object store
         self._save_state(state=state, name="vocab", access=access)
 
-    def load_pipeline(self, pipeline_template, states):
+    # Find argument type : pipeline_template and states
+    def load_pipeline(self, pipeline_template, states) -> None:
```
Comment (Nilanshrajput) [inline syfertext/language.py:130]:
```suggestion
    @type_hints
    def load_pipeline(self, pipeline_template: List[dict], states: Dict[str, dict]) -> None:
```

File: syfertext/language_model.py [hunk 5]
```diff
@@ -144,13 +149,15 @@ def send(self, location: BaseWorker) -> LanguageModelPointer:
...
+    @type_hints
     def create_pointer(
+        self,
         language_model: "LanguageModel" = None,
         owner: BaseWorker = None,
         location: BaseWorker = None,
         id_at_location: str = None,
         tags: Set[str] = None,
-        ptr_id = None,
+        ptr_id: str = None,
```
Comment (Nilanshrajput) [inline syfertext/language_model.py:160]:
```suggestion
        ptr_id: Union[str, int] = None,
```

File: syfertext/language_model.py [hunk 6]
```diff
@@ -144,13 +149,15 @@ def send(self, location: BaseWorker) -> LanguageModelPointer:
         return language_model_pointer
 
     
+    @type_hints
     def create_pointer(
+        self,
         language_model: "LanguageModel" = None,
         owner: BaseWorker = None,
         location: BaseWorker = None,
         id_at_location: str = None,
```
Comment (Nilanshrajput) [inline syfertext/language_model.py:158]:
```suggestion
        id_at_location: Union[str, int] = None,
```

File: syfertext/pipeline/simple_tagger.py [hunk 7]
```diff
@@ -142,6 +150,7 @@ def _get_tag(self, token: Token):
         return tag
 
     @staticmethod
+    # Find return type: Tuple of what?
     def simplify(worker: BaseWorker, simple_tagger: "SimpleTagger"):
```
Comment (Nilanshrajput) [inline syfertext/pipeline/simple_tagger.py:154]: tuple of str.Union[set, list, dict],object,bool

File: syfertext/pointers/state_pointer.py [hunk 8]
```diff
@@ -43,7 +46,9 @@ def __init__(
             garbage_collect_data=False,
         )
 
-    def get_copy(self) -> "State":
+    # On @type_hints gives : State not defined, on importing leads to a circular import
```
Comment (Nilanshrajput) [inline syfertext/pointers/state_pointer.py:49]: let's just go with strings cc @AlanAboudib

File: syfertext/span.py [hunk 9]
```diff
@@ -74,7 +79,9 @@ def set_attribute(self, name: str, value: object):
 
         setattr(self._, name, value)
 
-    def __getitem__(self, key: Union[int, slice]):
+    # Find return type: Union[token, span]
```
Comment (Nilanshrajput) [inline syfertext/span.py:82]: yup that's right return type

File: syfertext/span.py [hunk 10]
```diff
@@ -289,6 +302,7 @@ def create_pointer(
         if owner is None:
             owner = span.owner
 
+        """ Gives unexpected keyword argument """
```
Comment (Nilanshrajput) [inline syfertext/span.py:305]: can you give more details here?

File: syfertext/string_store.py [hunk 11]
```diff
@@ -35,7 +37,8 @@ def __contains__(self, string):
 
         return string in self.str_to_key.keys()
 
-    def add(self, string: str):
+    # Some issue with add method
+    def add(self, string:str) -> int:
```
Comment (Nilanshrajput) [inline syfertext/string_store.py:41]: is it related to type hints?

File: syfertext/string_store.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+#from .typecheck.typecheck import type_hints
```
Comment (Nilanshrajput) [inline syfertext/string_store.py:1]: import commented!

File: syfertext/string_store.py [hunk 13]
```diff
@@ -97,7 +101,8 @@ def __getitem__(self, string_or_id: Union[str, int]):
 
             return key
 
-    def __len__(self):
+
```
Comment (Nilanshrajput) [inline syfertext/string_store.py:104]:
```suggestion
    @type_hints
```

File: syfertext/token.py [hunk 14]
```diff
@@ -87,6 +93,7 @@ def remove_attribute(self, name: str):
 
         delattr(self._, name)
 
+    # Find return type : custom attribute
```
Comment (Nilanshrajput) [inline syfertext/token.py:96]: `object `if it doesnt work use `typing.Any`

File: syfertext/token.py [hunk 15]
```diff
@@ -192,7 +209,8 @@ def similarity(self, other):
 
         return sim
 
-    def get_encrypted_vector(self, *workers, crypto_provider=None, requires_grad=True):
+    # Find argument type : *workers
```
Comment (Nilanshrajput) [inline syfertext/token.py:212]: cc @AlanAboudib  any idea, how to handle this, maybe *BaseWorker :)

File: syfertext/string_store.py (hunk 12 above)
Comment (hershd23) [reply syfertext/string_store.py:1]: Yeah there was an issue with stringstore.py, so I have removed the import for now

File: syfertext/pipeline/simple_tagger.py (hunk 7 above)
Comment (hershd23) [reply syfertext/pipeline/simple_tagger.py:154]: I did not get that completely

File: syfertext/string_store.py (hunk 11 above)
Comment (hershd23) [reply syfertext/string_store.py:41]: Yeah, it has to do with typehints, I'll remove that

File: syfertext/span.py (hunk 10 above)
Comment (hershd23) [reply syfertext/span.py:305]: I did not get this error while running any tests, but my linter threw this error
Here is the definition of StatePointer, the init does not have the attribute type garbage_collect_data,
Might/Might not be an issue
```
class StatePointer(ObjectPointer):
    """This class defines a pointer to a State object. Whenever a State
    object is searched on the grid and found on any of the remote workers,
    a pointer to it, represented by this object, is returned. Then using
    this object, a copy of the state can be downloaded to the worker that
    requested it.
    """

    def __init__(
        self,
        location: BaseWorker,
        id_at_location: str,
        owner: BaseWorker,
    ):
        # Initialize the parent object
        super(StatePointer, self).__init__(
            location=location,
            id_at_location=id_at_location,
            owner=owner,
            id=None,
            garbage_collect_data=False,
        )
```
